main.tf

```diff
@@ -3,43 +3,51 @@ resource "hcp_vault_cluster" "main" {
-  #      #      grafana_password   = try(var.audit_log_config.grafana_password, null)
-  #      #      grafana_user       = try(var.audit_log_config.grafana_user, null)
-  #      #      splunk_hecendpoint = try(var.audit_log_config.splunk_hecendpoint, null)
-  #      #      splunk_token       = try(var.audit_log_config.splunk_token, null)
-  #    }
-  #  }
+  # see https://registry.terraform.io/providers/hashicorp/hcp/latest/docs/resources/vault_cluster#audit_log_config
+  # and https://developer.hashicorp.com/terraform/language/expressions/dynamic-blocks
+  dynamic "audit_log_config" {
+    # Metrics Configuration is not allowed on `dev` tier
+    for_each = ((!can(regex("^dev$", var.tier)) && var.audit_log_config.enabled) ? [1] : [])
+
+    content {
+      datadog_api_key    = try(var.audit_log_config.datadog_api_key, null)
+      datadog_region     = try(var.audit_log_config.datadog_region, "us1")
+      grafana_endpoint   = try(var.audit_log_config.grafana_endpoint, null)
+      grafana_password   = try(var.audit_log_config.grafana_password, null)
+      grafana_user       = try(var.audit_log_config.grafana_user, null)
+      splunk_hecendpoint = try(var.audit_log_config.splunk_hecendpoint, null)
+      splunk_token       = try(var.audit_log_config.splunk_token, null)
+    }
+  }
 
   # see https://registry.terraform.io/providers/hashicorp/hcp/latest/docs/resources/vault_cluster#nestedblock--major_version_upgrade_config
-  major_version_upgrade_config {
-    upgrade_type            = var.major_version_upgrade_config.upgrade_type
-    maintenance_window_day  = var.major_version_upgrade_config.maintenance_window_day
-    maintenance_window_time = var.major_version_upgrade_config.maintenance_window_time
+  dynamic "major_version_upgrade_config" {
+    # Major Version Upgrade configuration may only be set on clusters of STANDARD or PLUS tier
+    # see https://github.com/hashicorp/terraform-provider-hcp/search?&q=only+allowed+for+STANDARD+or+PLUS+clusters
+    for_each = (can(regex("^standard_|^plus_", var.tier)) ? [1] : [])
+
+    content {
+      upgrade_type            = var.major_version_upgrade_config.upgrade_type
+      maintenance_window_day  = var.major_version_upgrade_config.maintenance_window_day
+      maintenance_window_time = var.major_version_upgrade_config.maintenance_window_time
+    }
   }
 
   # see https://registry.terraform.io/providers/hashicorp/hcp/latest/docs/resources/vault_cluster#metrics_config
   # and https://learn.hashicorp.com/tutorials/cloud/vault-metrics-guide
-  # and https://www.terraform.io/language/expressions/dynamic-blocks
+  # and https://developer.hashicorp.com/terraform/language/expressions/dynamic-blocks
   dynamic "metrics_config" {
-    for_each = (var.metrics_config.enabled ? [1] : [])
+    # Metrics Configuration is not allowed on `dev` tier
+    for_each = ((!can(regex("^dev$", var.tier)) && var.metrics_config.enabled) ? [1] : [])
```

Comment:
Total judgement call here, but I prefer `? ["enabled"] : []`.

Simply put "enabled" is slightly more descriptive than "1".

Reply:
I like this - much clearer!